Issue 2719 - Map-maxi update

Fybex opened the pull request:
## Important
To merge the map, you must wait for merge #3517
For test, you need to install new dependencies: `npm ci`

# Description

### Map-maxi is ready 🎉 

Now we have two map providers:
- open street maps (by default)
- google maps
Both of them have the same functionality.

## Functionality
we can put multiply markers which has pop-ups on click, for add marker you can hold mouse on map until alert will appear so add marker, or use search box to put marker in special place, on marker click, pop-up will open, you can set title and description or remove marker.
Also, you can drag marker to change his placement.

On the frontend, map will be positioned as it was on last moment on editor, with zoom the same story.

<!---
Please include a summary of the changes and the related issue.
Please also include relevant motivation and context.
--->

Fixes #2719 

# How Has This Been Tested?
Checked if new map working good in general with both providers, on frontend and backend.

<!---
Please describe the tests that you ran to verify your changes.
Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration.
--->

# Test checklist

<!--- Please remove the unnecessary checkbox --->

**_ Front/Back Testing _**

-   [ ] Test the map-maxi open street maps map provider
-   [ ] Test the map-maxi Google Maps  map provider
-   [ ] Test the map-maxi all other functionality
-   [ ] Check that the settings work on frontend
-   [ ] Check that backend works and saves the settings after the editor reload
-   [ ] Same 1-4 points on responsive
-   [ ] Test the block inside the grid (Container + Row + Column)
-   [ ] Test the block as a standalone block
-   [ ] Duplicate the block, test that the settings of the first do not affect the second
-   [ ] Test with 2 blocks of the same type

**_ Pre-Code Testing _**

-   [ ] Test the map-maxi open street maps map provider
-   [ ] Test the map-maxi Google Maps  map provider
-   [ ] Test all other the map-maxi functionality
-   [ ] Check that the settings work on frontend
-   [ ] Check that backend works and saves the settings after the editor reload
-   [ ] Same 1-4 points on responsive
-   [ ] Test the block inside the grid (Container + Row + Column)
-   [ ] Test the block as a standalone block
-   [ ] Duplicate the block, test that the settings of the first do not affect the second
-   [ ] Test with 2 blocks of the same type
-   [ ] Check no commented code and no unnecessary imports
-   [ ] Standards of the project have been followed
-   [ ] No errors/warnings on console

# Checklist

-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my code
-   [X] I have commented my code, particularly in hard-to-understand areas
-   [X] My changes generate no new warnings/errors
-   [X] I have added/updated tests that prove my fix is effective or that my feature works
-   [X] New and existing unit tests pass locally with my changes

Fybex committed:
Add dependencies for maxi-map

Fybex committed:
issue-2719 map-maxi upgrade

Fybex committed:
Dragging marker, adding by holding

Fybex committed:
Add popups to map-maxi icons

Fybex committed:
Prefix support

Fybex committed:
Attributes update

Fybex committed:
Styles update

Fybex committed:
inspector update

- map-marker upgrade
- add marker pop-up settings
- add map interaction settings

Fybex committed:
Update edit.js map-maxi

Fybex committed:
Finished osm backend map-maxi

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
remove autoFocus from searchbox button

Fybex committed:
Close searchbox results when add marker

Fybex committed:
Fix styles map-maxi

Fybex committed:
temporary add browserslist

elzadj committed:
package

Naaaaiix committed:
Update dependencies

Naaaaiix committed:
Resolve compiler conflict

Naaaaiix committed:
Update node version on Github actions

Naaaaiix committed:
Fix unit test dependency lack

Naaaaiix committed:
Fix unit test

Naaaaiix committed:
Fix test

Naaaaiix committed:
Merge branch 'master' into issue-2756-dependencies

Naaaaiix committed:
Skiped non-update tests

Naaaaiix committed:
Merge branch 'master' into issue-2756-dependencies

Naaaaiix committed:
Fix conflicts on webpack config file

Naaaaiix committed:
Merge branch 'issue-2719' into map-with-updated-dependencies

Fybex committed:
update react-leaflet

Fybex committed:
Merge branch 'master' into issue-2719

Naaaaiix committed:
Update dependencies

Naaaaiix committed:
Fixes postcss and autoprefixer

elzadj committed:
Issue 2756 dependencies css (#2892)

* Add MiniCssExtractPlugin

* Remove comments from built css

* Update webpack.config.js

Co-authored-by: Josh <50477222+Naaaaiix@users.noreply.github.com>

Co-authored-by: Josh <jjuancap@gmail.com>
Co-authored-by: Josh <50477222+Naaaaiix@users.noreply.github.com>

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Merge branch 'issue-2756-dependencies' into issue-2719

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Map-maxi frontend sketch

Fybex committed:
Fix prefix unit tests failing

Fybex committed:
Merge branch 'master' into issue-2719

elzadj committed:
Fix react-leaflet

elzadj committed:
Fix conflicts

elzadj committed:
Merge branch 'master' of https://github.com/yeahcan/maxi-blocks into issue-2719

elzadj committed:
Some fixes

elzadj committed:
Merge branch 'master' of https://github.com/yeahcan/maxi-blocks into issue-2719

elzadj committed:
Fix map ref

Fybex committed:
defaultMarkers default export

Fybex committed:
findMarkers refactor

promises -> async/await

Fybex committed:
Remove deprecated HOC

Fybex committed:
Map-control SVG fixes

Fybex committed:
Inspector fixes

- move inspectorTabs.blockSettings outside of settings accordion
- replace parentBlockStyle with blockStyle
- remove changeSVGContent

Fybex committed:
Styles update

- remove getCustomFormatsStyles
- replace parentBlockStyle with blockStyle

Fybex committed:
min/max Zoom instant check

Fybex committed:
Create Markers separate component

Fybex committed:
Replace ref by using useMap react-leaflet hook

Fybex committed:
Block description upd

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Remove unnecessary import

Fybex committed:
Improve disabled logic

Fybex committed:
Transform dragging and adding attrs to states

Fybex committed:
Markers optimize add logic, remove default value for markers attribute

Fybex committed:
Minor fixes

- fix isFirstClick word
- structure some props

Fybex committed:
Add google api key to maxiMap customData

Fybex committed:
Remove @googlemaps/js-api-loader, add react-leaflet-google-layer

Fybex committed:
Add GoogleMutant to maxi-map frontend

Fybex committed:
Map markers refactor

- create`mapMarker` attributes group
- fix `getSVGStyles` in map-maxi `styles.js`
- remove marker opacity settings
- fix marker width responsive settings

Fybex committed:
mapPopup attributes group

Fybex committed:
Fix marker, popup presets active status

Fybex committed:
Give correct and only needed props to inspector controls

Fybex committed:
Map controls separation

Fybex committed:
mapInteraction attributes group

Fybex committed:
MapPopupTextControl

Fybex committed:
MapControl fix

Fybex committed:
Change class to BEM methodology

+ fix `renderToString`

Fybex committed:
Minor fixes

- add classes to some instances
- fix leaflet issue on remove marker
- fix first click prevent showing alert
- change classnames

Fybex committed:
More fixes

- Marker remove button: dataindex -> data-index
- saving correct coords

Fybex committed:
Move map controls to components in map-maxi folder

Fybex committed:
Basic map e2e tests

Fybex committed:
PresetsControl

Fybex committed:
Map markers width responsive indicators fix

Fybex committed:
Small refactor

Fybex committed:
Map attributes alphabet + height fix

Fybex committed:
Minor fixes

- add relation tab
- remove `getMapStyles`
- fix some things in `styles.js`

Fybex committed:
Frontend description class fix

Fybex committed:
getSVGWidthStyles transform props to obj

Fybex committed:
Optimize and move `orderAttributes` to top

Fybex committed:
`getPrefix` func

Fybex committed:
Implement ability to get `prefix` from `copy-paste-mapping.js` objs

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Merge branch 'issue-3496' into issue-2719

Fybex committed:
Refactored code for new copypaste schema

Fybex committed:
Schema minor change

Fybex committed:
Remove `orderAttributes`

Fybex committed:
Updated button copy-paste

Fybex committed:
Advanced template

Fybex committed:
Schema logic change

add group prop to define groups and conditions

Fybex committed:
column-maxi copy-paste

Fybex committed:
container-maxi copy-paste

Fybex committed:
divider-maxi copy-paste

Fybex committed:
group-maxi copy-paste

Fybex committed:
image-maxi copy-paste

Fybex committed:
number-counter copy-paste

Fybex committed:
row copy-paste

Fybex committed:
Remove unnecessary code since #3191

Fybex committed:
svg copy-paste

Fybex committed:
text copy-paste

Fybex committed:
video copy-paste

Fybex committed:
Fix checkbox selection on same labels for group label and nested label

Fybex committed:
Remove template value on template add

Fybex committed:
Add + update unit copy-paste tests

Fybex committed:
Update e2e

Fybex committed:
Minor fixes

- prevent adding `group` in specialPaste
- not reset group checkbox on tab switch
- make specialPaste defaults not that specific

Fybex committed:
Remove comment

Fybex committed:
Sentence case

Fybex committed:
Fix different font-size for group items and not on the highest level

Fybex committed:
Simplify schema, add ability to have array of props without object

Fybex committed:
Add `normalizeLabel`

Fybex committed:
Fix size template

Fybex committed:
Fix and add e2e

Fybex committed:
Remove temporary code

Fybex committed:
Minor copy-paste blocks obj fixes

Fybex committed:
Unit tests update

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Merge branch 'issue-3496' into issue-2719

Fybex committed:
Map copy-paste

Fybex committed:
Minor fixes

Fybex committed:
Update popup presets

Fybex committed:
Make `styles.js` code easier

Fybex committed:
Popup border width min/max settings

Fybex committed:
Container fixes

Fybex committed:
Big cleanup

Fybex committed:
Add popup close button

Fybex committed:
Show `InfoBox` only when provider is googlemaps

Fybex committed:
Remove removed from copypaste

Fybex committed:
Fix grey half map tiles

Fybex committed:
Make no api key link correct

Fybex committed:
Divide `edit.js` to components

Fybex committed:
Remove not used variable

Fybex committed:
Popup typography fix

Fybex committed:
Fix e2e tests

- delete deprecated map-control
- fix svg width unnecessary added scape
- update map snapshots

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Split `mapPopup` group to `mapPopup` and `mapPopupText`

Fybex committed:
Fix default typography

Fybex committed:
Fix e2e

Fybex committed:
CustomCss map update

Fybex committed:
Transitions

Fybex committed:
Update e2e

Fybex committed:
Fix e2e

Fybex committed:
Replace `waitForTimeout` with `waitForSelector`

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Update dependencies

Marko64 reviewed (approved):
Worked wonderfully. Going to approve, but I do have a few small comments:

1. Not a big deal if we leave it as is, but the _min_ can be greater than the _max_ zoom level.
![Screenshot_4](https://user-images.githubusercontent.com/19425503/182574897-e0c7f6e5-157b-48e7-99df-04ffb0030663.jpg)

2. When you enter the wrong Google maps API key, the map is just grey. Is it possible to show a warning message in this case? Something like what Google says in the console: "Invalid map API key".

3. Can't increase width above 600px or similar. 
Set width to fit content may not be needed for the map.

4. Interactions:
Dragging works on the frontend only. Other settings work on the backend too. Not a big deal as well, but mentioning just in case.

Fybex committed:
Create `get_block_data` func

Fybex committed:
Apply @TeksPeks suggestions

Fybex committed:
Other minor fixes

Fybex committed:
Update e2e

Naaaaiix reviewed (commented):
I'm running out of sentences saying how great is a PR and implementation, so going to start using gifs:
![](https://media.giphy.com/media/czubJ08i7deuKGJE9A/giphy.gif)

Some comments, but in general is a 10. Congrats @Fybex 🎉 

- On the `components` folder inside of blocks replace the folder for files directly. We don't need, for example, a path like `src/blocks/map-maxi/components/search-box/index.js`; would be better, IMO, to have something like `src/blocks/map-maxi/components/search-box.js` 👍

Fybex committed:
Simplify `getMapContainer`

Fybex committed:
`MapMarkersControl` refactor + custom color fix

Fybex committed:
Make `getSVGStyles` work with map markers

Fybex committed:
Remove unnecessary return

Fybex committed:
:fire: svg fill opacity

Fybex committed:
Get rid of `mapRef` + add comment

Fybex committed:
Remove ternary operator

Fybex committed:
:fire: `PresetsControl`

replace by `DefaultStylesControl`

Fybex committed:
Dark magic

remove `!important`, trying to adjust popup position based on marker size(can't attach a mathematical explanation, I don't understand it myself)

Fybex committed:
Update e2e

Fybex committed:
Merge branch 'master' into issue-2719

Fybex committed:
Fix package-lock

Naaaaiix reviewed (approved):
Let's fix these e2e tests and merge this beauty 😜
